### DISCUSSION TITLE
Optimisers

### DISCUSSION THREAD

**Discussion by jbcaillau** [control-toolbox/control-toolbox.github.io#47] [Ideas]
@ocots @frapac @joseph-gergaud @PierreMartinon 

> Bonjour Olivier, Jean-Baptiste,
>
> et merci pour ce retour ! Je pense qu'on se complémente bien en effet : nous sommes de notre coté justement intéressé par avoir une base de benchmark.

Parfait

> Pour la suite, je pense que le plus simple est que je prenne en main OptimalControl.jl et essaie d'utiliser MadNLP pour > résoudre les problèmes, pour voir. À moyen terme, je me demande si nous pourrions essayer de monter un stage en > commun pour essayer d'implémenter un solveur KKT dédié aux problèmes de controle optimal, ala Stephen Wright.

Oui : OptimalControl.jl peut être simplement utilisé comme une interface de modélisation ; que ce soit avec la syntaxe "abstraite" [^1] ou "fonctionnelle" [^2], le pb de contrôle optimal est traduit
- tout d'abord en une liste de contraintes NLP sous la forme de contraintes linéaires élémentaires (boites sur un range d'indices) ou non-linéaires (boites sur des fonctions), voir [nlp_constraints](https://control-toolbox.org/CTBase.jl/dev/api-model.html#CTBase.nlp_constraints-Tuple{OptimalControlModel})
- ensuite sous la forme d'un ADNLPModel ; cette traduction est pour l'instant interne à la fonction [CTDirect.solve](https://control-toolbox.org/CTDirect.jl/dev/api.html#CTDirect.solve-Tuple{OptimalControlModel,%20Vararg{Any}}) mais a vocation à être extraite et généralisée à d'autres modeleurs / solvers. Voir [cette issue](https://github.com/control-toolbox/CTDirect.jl/issues/70).

Volontiers pour dév. commun, par exemple en s'appuyant sur un stage. Calons un point ensemble pour préciser.

[^1]: voir [cet exemple](https://control-toolbox.org/docs/optimalcontrol/dev/tutorial-basic-example-f.html)

[^2]: voir [cet autre exemple](https://control-toolbox.org/docs/optimalcontrol/dev/tutorial-basic-example.html)

**Comment by ocots** [comment]
Dans un premier temps, on aimerait avoir le problème de Goddard et d'autres cas d'usage sous les différentes formes[^1] données par le diagramme ci-dessous.

> **Légende.** Quand je mets `.jl` dans le diagramme c'est que l'on utilise le package mais je ne détaille pas comment. Sinon, je parle directement du type. `OCP` c'est le problème de contrôle optimal donné par sa formulation mathématique.

```mermaid
stateDiagram-v2
	OCP --> JuMP.jl
	OCP --> ExaModels.jl
	OCP --> OptimalControl.jl
	OptimalControl.jl --> ADNLPModel
	JuMP.jl --> Ipopt.jl
	ExaModels.jl --> ExaModel
	ADNLPModel --> AbstractNLPModel 
        ExaModel --> AbstractNLPModel 
        AbstractNLPModel --> NLPModelsIpopt.jl
	AbstractNLPModel --> MadNLP.jl
        NLPModelsIpopt.jl --> Ipopt.jl
```

- [ADNLPModels.jl](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl#adnlpmodels) : depuis `OptimalControl.jl`, on définit un `ADNLPModel` qui est un sous-type de `AbstractNLPModel`.
- [ExaModels.jl](https://github.com/sshin23/ExaModels.jl) : le type `ExaModel` est un sous-type de `AbstractNLPModel`. Pour qu'un solveur puisse gérer un nouveau type, sous-type de `AbstractNLPModel`, il suffit de surcharger quelques fonctions de `NLPModels.jl` qui retourne la fonction, son gradient... C'est le multiple dispatch qui gère le truc. En effet, puisqu'il n'y a pas d'héritage, on passe par le multiple dispatch.
- [Ipopt.jl](https://github.com/jump-dev/Ipopt.jl#ipoptjl)
- [JuMP.jl](https://github.com/jump-dev/JuMP.jl)
- [MadNLP.jl](https://github.com/MadNLP/MadNLP.jl) : il peut résoudre un problème défini par un `AbstractNLPModel` donc un `ExaModel` ou un `ADNLPModel` par exemple. **Remarque.** Il peut aussi se brancher à `JuMP` et donc résoudre un problème défini par `JuMP`.
- [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl) : défini `AbstractNLPModel`.
- [NLPModelsIpopt.jl](https://github.com/JuliaSmoothOptimizers/NLPModelsIpopt.jl) : ce n'est pas pour modéliser mais c'est un wrapper d'ipopt capable de résoudre un problème défini par un `NLPModel`.
- [OptimalControl.jl](https://github.com/control-toolbox/OptimalControl.jl)

[^1]: vous pouvez faire des modifications sur le diagramme.

**Comment by ocots** [comment]
Voulons-nous : 

```mermaid
flowchart TB
B(OCP) --> OptimalControl.jl --> Packages
Types --> AbstractNLPModel
AbstractNLPModel --> Wrappers
subgraph Modelisation
direction TB
subgraph Packages
ADNLPModels.jl
ExaModels.jl
JuMP.jl
end
subgraph Types
ADNLPModel
ExaModel
NLPModelsJuMP
end
ADNLPModels.jl --> ADNLPModel
ExaModels.jl --> ExaModel
JuMP.jl --> NLPModelsJuMP
AbstractNLPModel
end
subgraph Resolution
subgraph Wrappers
NLPModelsIpopt.jl
NLPModelsKnitro.jl
end
subgraph Solvers
Ipopt.jl
KNITRO.jl
MadNLP.jl
end
NLPModelsIpopt.jl --> Ipopt.jl
NLPModelsKnitro.jl --> KNITRO.jl
end
AbstractNLPModel --> MadNLP.jl
JuMP.jl --> Solvers
```

- [KNITRO.jl](https://github.com/jump-dev/KNITRO.jl) : **Remarque.** `KNITRO` peut aussi se brancher sur `JuMP` bien entendu.
- [NLPModelsKnitro.jl](https://github.com/JuliaSmoothOptimizers/NLPModelsKnitro.jl) : c'est un wrapper à `KNITRO` pour résoudre un problème défini par un `AbstractNLPModel`.
- [NLPModelsJuMP](https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl#nlpmodelsjump)

**Reply by ocots** [reply]
Ce serait bien d'avoir un benchmark de ces méthodes sur une liste de problèmes de contrôle optimal.

**Reply by jbcaillau** [reply]
@ocots Yes! Going towards a list of control problems (*not* their discretisation as math. programs or so) using a high level description (*e.g.* `OptimalControl` abstract syntax 🙂) would be a great outcome of [CTProblems](https://github.com/control-toolbox/CTProblems.jl)

**Reply by ocots** [reply]
We need to define an `AbstractOCPModel` or `AbstractOptimalControlModel` and methods on it, used by our solvers and functors such as `CTDirect`. If everything is written taking as argument an `AbstractOCPModel`, then it becomes easier to implement different OCP modelisations. Then, from our `AbstractOCPModel` we get an NLP thanks for instance to `CTDirect`:

```julia
using OptimalControl
ocp = Model()      # ocp is a OCPModel <: AbstractOCPModel
nlp = Direct(ocp, :adnlp) # nlp is a AbstractNLPModel
```

or

```julia
ocp = Model()
ocp = Model(:jump) # we define the model à la JuMP?
ocp = Model(:exa) # ...
```

and 

```julia
nlp = Direct(ocp, :adnlp)
nlp = Direct(ocp, :jump)
nlp = Direct(ocp, :exa)
```

**Reply by ocots** [reply]
Maybe we need also a type: `AbstractDOCPModel` for abstract discretized optimal control problem. 

```julia
using OptimalControl
ocp = Model()
docp = Direct(ocp, steps=100, scheme=:trapeze)
nlp = NLP(docp, :adnlp) # or nlp = NLP(docp, :exa)
typeof(nlp) <: AbstractNLPModel # return true
```

and we can solve `nlp` by any solver we want. Other possibilities:

```julia
using OptimalControl
ocp = Model()
docp = Direct(ocp, steps=100, scheme=:trapeze)
solve(docp, :adnlp, :ipopt)
```

or 

```julia
using OptimalControl
ocp = Model()
solve(ocp, :direct, :adnlp, :ipopt, steps=100, scheme=:trapeze)
```

**Reply by PierreMartinon** [reply]
This could be a good idea. As a side note, when writing the transcription function (currently takes an ocp ans returns the corresponding nlp) I encountered the small problem of the initial guess: ADnlp typically expects the starting point to be given, while for me this is not strictly part of the discretized problem but more a piece of information passed to the solve() function.

Back to your first example block, we could for instance decide that the initial guess would be normally passed to the NLP() function, not the Direct() function. More generally, I feel it should be possible to call solve() with a specific initial point, so we would need a function that would set/update the initial point in the NLP problem anyway. Worst case scenario, we'll need to create a new nlp for each initial guess.

**Comment by ocots** [comment]
On remarque deux choses : 

- Pour `NLPModels` ils font eux-mêmes les wrappers aux solveurs là où pour `JuMP`, ce sont les solveurs qui font en sorte d'être compatibles en passant par `MathOptInterface`.
- En ce qui concerne `MadNLP`, ils se sont rendus compatibles à la fois à `JuMP` et à `NLPModels`.

**Comment by jbcaillau** [comment]
@frapac @ocots @PierreMartinon @joseph-gergaud 

- [MadNLP.jl](https://madnlp.github.io/MadNLP.jl/dev): perf. *vs.* Ipopt : approx. x1.2, interface with [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl)
- [ExaModels.jl](https://github.com/sshin23/ExaModels.jl): no user defined functions (as JuMP), but allows vector constraints (JuMP doesn't)
- for the record: [ADNLPModels.jl](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl) can use several backends ([ForwardDiff.jl](https://juliadiff.org/ForwardDiff.jl/stable/), in practice) for AD of full Julia code
- motivation to allow full Julia expressions: avoid code duplication between several DSLs (the common base should be Julia not yet another DSL)
- done in [OptimalControl.jl](https://github.com/control-toolbox/OptimalControl.jl), including no "world age pb" (check [this test](https://github.com/control-toolbox/CTBase.jl/blob/ab0637989862bba25158d6fd4b80faa0bd9b4aa8/test/test_onepass.jl#L116))
- idea: allow ExaModels expressions (it is Julia code, indeed) inside OptimalControl `@def`? (Implies a specific post-treatment)
- [x] @frapac @ocots internship (M1/M2 level): use case ([Goddard](https://ct.gitlabpages.inria.fr/gallery/goddard-j/goddard.html) and more) to compare JuMP  and MadNLP
- related topic: test with [Knitro.jl](https://github.com/jump-dev/KNITRO.jl) (*cf.* trust-region *vs.* interior point)
- conferences mentioned by @gergaud: [SMAI-MODE](https://indico.math.cnrs.fr/event/9418/page/630-accueil) (deadline 18/12/2023, no MS), [FGS'2024](https://www.unioviedo.es/fgs2024/) (MS deadline 14/02/2024)
- [x] @frapac please send a pointer towards the power grid optim. benchmark (Los Alamos) you indicated

[^1]: Currently, the only stable / complete / robust AD backend is ForwardDiff (*vs.* Zygote, *etc.*) In particular, no completely satisfying tool for reverse mode. Ongoing: [Enzyme.jl](xxxx)

**Reply by frapac** [reply]
the benchmark of the different AD backends on the ACOPF problems can be found here: https://discourse.julialang.org/t/ac-optimal-power-flow-in-various-nonlinear-optimization-frameworks/78486/59

We should keep in mind that JuMP remains slower than AMPL by a factor of two. In that regards, ADNLPModels.jl is not tractable for us (yet).

**Reply by frapac** [reply]
has someone tried [FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl)? The original package is around since the 2000s, and has been ported in Julia recently. I would be curious to know the performance of that other AD backend.

**Reply by jbcaillau** [reply]
@frapac not heard of it, having a look (-> added in [this issue](https://github.com/control-toolbox/CTBase.jl/issues/25#issuecomment-1774504580))
> has someone tried [FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl)? The original package is around since the 2000s, and has been ported in Julia recently. I would be curious to know the performance of that other AD backend.

**Comment by jbcaillau** [comment]
PS. Congrats @frapac and co-authors !!!

https://www.linkedin.com/posts/fran%C3%A7ois-pacaud-99b01487_github-madnlpmadnlpjl-a-solver-for-nonlinear-activity-7120440464666701825-_evI?utm_source=share&utm_medium=member_ios

**Reply by frapac** [reply]
thank you :-) 

I am glad MadNLP and ExaModels got some recognition from COIN-OR. I hope this will give some momentum to the MadNLP ecosystem!

**Comment by ocots** [comment]
Ideas for an internship:

In a first place it could be great to compare some pairs of (model, solve) for some case studies: Goddard... See the list of problems [here](https://control-toolbox.org/docs/ctproblems/stable/problems-list.html).

The list of pairs of models and solvers (to be completed, updated) :

| Model              | Solver      |
| ---                   | ---            | 
| JumP               | Ipopt        |
| JumP               | KNITRO   |
| JumP               | MadNLP  |
| ADNLPModel  | Ipopt       |
| ADNLPModel  | KNITRO  |
| ADNLPModel  | MadNLP |
| ExaModel        | Ipopt       |
| ExaModel        | KNITRO  |
| ExaModel        | MadNLP |
| Casadi             | Ipopt       |
| Casadi             | KNITRO  |
| Bocop              | Ipopt      |

Then, it could be great to define a process for benchmarking with online results and with the possibility for users to make their own benchmarks.

**Reply by ocots** [reply]
We need to define in which terms we make the comparisons.  We should also compare the cost of a call to the derivatives.

**Reply by ocots** [reply]
WORHP ?

**Reply by ocots** [reply]
Pour le choix des problèmes, voir benchmark [COPS](https://www.mcs.anl.gov/~more/cops/) et son implémentation JumP, cf. package [COPSBenchmark.jl](https://github.com/MadNLP/COPSBenchmark.jl).

**Reply by ocots** [reply]
Les différents problèmes intéressants : Goddard, un problème lisse, du chattering, de nombreuses commutations (transfert orbital), contraintes d'état. Contraintes [MPEC](https://arxiv.org/abs/2312.11022), problèmes hybrides (au delà du stage).

**Comment by jbcaillau** [comment]
@ocots @PierreMartinon @joseph-gergaud check [this post](https://discourse.julialang.org/t/jump-constraint-with-a-huge-expression/108456/24?u=jbcaillau) on JuMP and related things (thanks @tmigot for pointing to ct). Most notably, [function tracing](https://jump.dev/JuMP.jl/stable/manual/nonlinear/#Function-tracing) now available in JuMP: @tmigot @amontoison does this entail full AD support (though not sparse, I guess) for the expressions generated by overloading on JuMP abstract expression type?

**Reply by amontoison** [reply]
I didn't find time yet @jbcaillau to check the new NLP API of JuMP.
Maybe this summer for the JuMP-dev at Montréal.

**Comment by ocots** [comment]
Stage M1 2024 : 

Livrables : 

- liste des problèmes tests dans un dépôt Github : 
   - Priorité 1 : définis sous JumP et OptimalControl.jl 
   - Priorité 2 : Casadi
   - Priorité 3 : Bocop (voir avec Pierre)
- performance profile et tableau instance vs comparaison des outils (temps dans la diff auto, temps total, nombre itérations, sparse pattern) sur les exemples tests pour les paires (Model, Solver) suivantes : 
   - Task 1 : (JuMP, Ipopt, JuMPDefault backend, _linear_solver), _linear_solver in {MUMPS, HSLMA57}
   - Task 2 : (JuMP, Ipopt, _backend, best_linear_solver), _backend in {JuMPDefault, [SymbolicAD](https://github.com/odow/SymbolicAD.jl), ExaModel}
   - Task 3 : (JuMP, _solver, best_backend, best_linear_solver), _solver in {Ipopt, Knitro_IPM, Knitro_SQP, MadNLP}
   - Task 4 : (ADNLPModel, Ipopt, best_linear_solver) : via OptimalControl.jl cf. remarque
   - Task 5 : OptimalControl.jl -> JuMP (*cf.* [point plus haut](https://github.com/orgs/control-toolbox/discussions/47#discussioncomment-8362362))

**Remarque :** ADNLPModel est utilisé via OptimalControl.jl. Pour le moment OptimalControl.jl ne fournit que ADNLPModel + Ipopt. 
   
Perspectives / Bonus : 

- ADNLPModel avec les solveurs Knitro et MadNLP si intéressant
- Voir Casadi.jl 
- Attention aux schémas de discrétisation, pour le moment seul les trapèzes sont disponibles dans la control-toolbox.

**Reply by jbcaillau** [reply]
@ocots @frapac trouvé un stagiaire M1 / M2 sur cette action opti ?

**Reply by frapac** [reply]
Nous en avions un mais il nous a fait faux bond. Aucun autre candidat pour le moment.

**Reply by ocots** [reply]
On devrait avoir quelqu'un ! Je vous tiens au courant.

**Comment by jbcaillau** [comment]
@amontoison @tmigot rappel des infos sur `JuMP` discutées dans [ce post plus haut](https://github.com/orgs/control-toolbox/discussions/47#discussioncomment-8362362) avec
- [function tracing](https://jump.dev/JuMP.jl/stable/manual/nonlinear/#Function-tracing) (essentiellement surcharge des opérateurs classiques, puis `ReverseAD` sur les expressions générées dans le DSL `JuMP`)
- [user defined operators](https://jump.dev/JuMP.jl/stable/manual/nonlinear/#jump_user_defined_operators) (and falling back to `ForwardDiff.jl`)

**Reply by amontoison** [reply]
Merci @jbcaillau!
Je comprends comment ça fonctionne maintenant.
Ça ne doit pas être bien compliqué à interfacer dans `NLPModelsJuMP.jl`.

D'un point de vue utilisateur, il faut néanmoins définir des fonctions qui peuvent prendre un `Vector{VariableRef}` comme argument.
C'est ce qu'on a fait avec Tangi dans ADNLPModels.jl.
La différence c'est qu'on utilise Symbolics.jl et qu'on peut définir `f(x::Vector{Float64})`.
Symbolics.jl arrive à nous donner le "sparsity pattern" du Jacobien / Hessien.

Je pense que JuMP arrive à faire une analyse plus en détails et isoler l'arbre d'expression de `f`.
Je vais en parler à Oscar à l'occasion.

**Reply by jbcaillau** [reply]
Le truc c'est de pouvoir utiliser des fonctions définies en julia standard, et non pas nécessairement des expressions bâties sur les briques de base de JuMP. Si leur surcharge (*aka* "function tracing") fait le job de façon raisonnablement large c'est OK. À voir.

**Comment by ocots** [comment]
@frapac For the benchmark we should have a look to:

- https://jso.dev/OptimizationProblems.jl/stable/
- https://github.com/JuliaSmoothOptimizers/SolverBenchmark.jl

**Comment by ocots** [comment]
AD benchmark : https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/issues/241

**Comment by amontoison** [comment]
L'équipe de contrôle optimal, vous utilisez quoi comme backend pour le Hessian creux dans ADNLPModels.jl?
Je viens d'y penser ce soir en répondant à un autre message, vous utilisez `SparseADHessian` ou `SparseReverseADHessian`?
Le second fait du "forward over reverse" et devrait être optimal si ReverseDiff arrive à manger votre fonction.
Par défaut c'est l'autre backend `SparseADHessian` qui fait du "forward over forward". Il est plus robuste parce que ForwardDiff.jl mange tout.

**Reply by ocots** [reply]
@PierreMartinon Une idée ?

**Reply by PierreMartinon** [reply]
Pour le moment on utilise l'option par defaut, donc *SparseADHessian* a priori, mais on pourrait tester le reverse oui !
On pourrait passer le choix de backend a *solve*/*directTranscription* (qui genere le ADNLPModel) et aussi avoir une fonction qui modifie un probleme existant du genre setADbackend(docp, ...), comme pour l'initialization.

Edit: ah en fait on utilise backend=:optimized

**Reply by jbcaillau** [reply]
merci @amontoison sur ce sujet, voir [ici](https://github.com/control-toolbox/CTBase.jl/issues/25#issuecomment-2182318402) : 
- il y a bien un `set_AD_backend`, voir [ici](https://github.com/control-toolbox/CTBase.jl/blob/3f144ca4f95a6fbc9bf2106c8d0e0d90d16eea7c/src/default.jl#L4)
- l'idée étant par ex. de pouvoir faire du reverse over forward 🤞🏽

@PierreMartinon just added this as an [issue](https://github.com/control-toolbox/CTDirect.jl/issues/117)

**Reply by ocots** [reply]
@frapac @0Yassine0 : Hessienne.
